### PR TITLE
partial fix #296416: accessibility of popup menu

### PR DESCRIPTION
**Pull request by MarcSabatella**
Resolves (in part): https://musescore.org/en/node/296416

Windows defines a menu key to allow context menus to be operated by keyboard.
We are triggering the popup menu in this case, but getting the wrong element:
we should be using the selection, not the element near the mouse.
This changes ScoreView::contextMenuEvent() to check for keyboard as reason,
and set the element appropriately if so.

Currently, screenreader support for the context menu is still limited, however.  It seems that that menu items corresponding to actions that are also in the main menu will not read, I think because NVDA realizes the parent of the action is not the object with the focus.  I'd love to find a solution for that too, but am more concerned with making sure the menu is accessible at all, and in particular, it's the commands that are *not* on the main menu that are most in need of beng accessible, and this PR does accomplish that.

### CODE REVIEW

File: mscore/scoreview.cpp [hunk 1]
```diff
@@ -320,6 +320,15 @@ void ScoreView::objectPopup(const QPoint& pos, Element* obj)
       popup->addAction("Debugger")->setData("list");
 #endif
 
+#if 0
```
Comment (dmitrio95) [inline mscore/scoreview.cpp]: If this code is not enabled anyway, would it perhaps be better to add it in a separate commit?

Comment (MarcSabatella) [reply mscore/scoreview.cpp]: You mean just put a TODO comment here but leave out the code?  I like to add a rough outline of the expected code where possible when I add a TODO, and it seems Werner favored that approach too.  Or maybe you mean, just add the TODO and disabled code as a separate commit within this same PR?

However, in this case I am fine removing it, because it actually appears the problem is not ours, it is a Qt bug.  As per one of the comments in the issue, the failure of the screenreader to read the items on the popup menu that are duplicated in the main menu is because of https://bugreports.qt.io/browse/QTBUG-79858.  So I will just remove this section, which would have just been a workaround even if I had been able to get it to work.

Comment (dmitrio95) [reply mscore/scoreview.cpp]: Let's better remove this section then, the other part of this PR is valuable itself, and the information added here is anyway present in the issue trackers, both our and Qt's ones.

**Comment by MarcSabatella** [comment]
Updated as per comments.

File: mscore/events.cpp [hunk 2]
```diff
@@ -849,7 +849,11 @@ void ScoreView::contextMenuEvent(QContextMenuEvent* ev)
       QPoint gp          = ev->globalPos();
       editData.startMove = toLogical(ev->pos());
       editData.buttons   = Qt::NoButton;
-      Element* e         = elementNear(editData.startMove);
+      Element* e         = nullptr;
+      if (ev->reason() == QContextMenuEvent::Keyboard)
+            e = score()->selection().element();
```
Comment (dmitrio95) [inline mscore/events.cpp]: Also wouldn't this make it impossible to open context menu for range or list selection? Maybe these cases should be handled too.

Comment (MarcSabatella) [reply mscore/events.cpp]: Oh, because Selection::element() returns null in those cases?  Good point, here it wouldn't hurt to return *something*.  Although, it's much less important, there isn't much in those context menus that can't be accessed another way.  Maybe just split staff?  Anyhow, elementNear() won't help as the location data is not useful, but I'll see what makes sense.

Comment (dmitrio95) [reply mscore/events.cpp]: Well, at a closer look, this will just behave like it had behaved before, since [this branch](https://github.com/musescore/MuseScore/blob/7278beddfaf6bab151fbcc896e6ed01b185a6dce/mscore/events.cpp#L868-L881) of code would be executed. So there is no regression with this patch.

That branch could probably be improved though by trying to find a relevant measure with something like [`selection().findMeasure()`](https://github.com/musescore/MuseScore/blob/7278beddfaf6bab151fbcc896e6ed01b185a6dce/libmscore/select.h#L191) or [`selection().firstChordRest().findMeasure()`](https://github.com/musescore/MuseScore/blob/7278beddfaf6bab151fbcc896e6ed01b185a6dce/libmscore/select.h#L189) in case of keyboard-originated context menu event and, following the existing logic, showing measure menu or usual page menu depending on whether a relevant measure was found. Maybe it would be reasonable to add it to this PR as well? If not, then it can be merged as it is anyway.

Comment (MarcSabatella) [reply mscore/events.cpp]: Not sure what kind of timetable you are looking at.  I can't guarantee I'd get to this today or tomorrow (although I might).  But I'd like to make the change in any case, it's a good improvement.  So whether this is merged now or later, I would want to do that.

Comment (dmitrio95) [reply mscore/events.cpp]: Actually the current plan is to finish merging changes for 3.3.3 release on Saturday, so if you add it until then it would perfectly fit into the current timeline. If not, then I would just merge this PR as it is, it is already an improvement anyway. So there is no hurry here in any case.

**Comment by MarcSabatella** [comment]
Updated to use findMeasure().  There's no completely ideal solution, really we eventually should figure out a way to do things like Select / All Similar Elements, which depends on mouse location and can't be faked by looking at the selection.  That is, we'd need a way to navigate while still having a range selection.  But anyhow, this was already an improvement, now it's somewhat more so.

File: mscore/events.cpp [hunk 3]
```diff
@@ -849,7 +849,14 @@ void ScoreView::contextMenuEvent(QContextMenuEvent* ev)
       QPoint gp          = ev->globalPos();
       editData.startMove = toLogical(ev->pos());
       editData.buttons   = Qt::NoButton;
-      Element* e         = elementNear(editData.startMove);
+      Element* e         = nullptr;
+      if (ev->reason() == QContextMenuEvent::Keyboard) {
+            e = score()->selection().element();
+            if (!e)
+                  e = score()->selection().findMeasure();
```
Comment (dmitrio95) [inline mscore/events.cpp]: It would probably be better to put handling of this case to ["else" branch](https://github.com/musescore/MuseScore/blob/7278beddfaf6bab151fbcc896e6ed01b185a6dce/mscore/events.cpp#L868-L881) since it would call `measurePopup()` rather than `objectPopup()` in this case, which seems to be more suitable. This is actually what I wrote about in my comment above.

File: mscore/scoreview.cpp (hunk 1 above)
Comment (dmitrio95) [reply mscore/scoreview.cpp]: This disabled code seems to have returned with the latest push to this branch.

File: mscore/events.cpp (hunk 3 above)
Comment (MarcSabatella) [reply mscore/events.cpp]: I see now.  I was thinking the goal was to only change behavior for the keyboard case, but you're actually talking about changing the behavior of right-click any time there is no nearby element, to try to use the selection.  That makes sense.

It's possible, I suppose, someone might prefer the current behavior where right-clicking an empty area of the score always give you access to the style & page settings dialogs, but I think the proposed behavior does make more sense, and you can always get the old behavior back by pressing Esc first.

I will update, and re-remove the disabled code (snuck back in because I was developing & testing on two different machines).

**Comment by MarcSabatella** [comment]
After some experimentation, I ended up not liking any of the alternatives I tried with respect to actual right-click behavior.  The main problem is, it matters which measure gets chosen, because measure properties cares, and it felt wrong having the chosen measure potentially be nowhere near the mouse cursor, givewn there are many cases where the mouse cursor actually seems like it should be considered within a measure (between the staves of a system, for instance).  And yet, if I relaxed the condition of the mouse point being within the staff, suddenly it's almost impossible to get to the page menu.  Maybe that's OK, but that's not a change I want to make right now.

If someone wants to make it in the future, just remove the two lines of code with the comment "but only use it if mouse pointer is within the staff".

**Comment by MarcSabatella** [comment]
So the current update only changes the behavior of the menu button and doesn't affect right-click at all.  But I did move the relevant code to the proper "else" clause (if not single element is identified) - I just check reason() again.

**Comment by MarcSabatella** [comment]
Which, I now see, is what you actually suggested to begin with.  Somehow I just wasn't getting it, but I think I do now, and I think what I have here is it!
